feat: wire node group for declarative validation

darshansreenivas opened the pull request:
/kind feature
/sig api-machinery
/sig storage

## What this PR does / why we need it

- Enables the **validation-gen** framework for the `node.k8s.io` API group (`v1beta1`, `v1`).
- Adds **Declarative Validation (DV)** support for **RuntimeClass**.
- Maintains parity with existing hand-written strategy validation as part of the migration.

## Which issue(s) this PR is related to

- Part of kubernetes/kubernetes#134280

## Special notes for your reviewer

- Scope is limited to enabling validation-gen and DV for `RuntimeClass`; no user-visible behavior changes intended.
- Includes unit tests that verify parity between declarative and hand-written validations (create/update paths).
- Follow-ups will extend DV to additional types in `node.k8s.io` as needed.

## Test plan

- Table-driven unit tests covering:
  - Create & update for `v1beta1` and `v1`
  - Required/immutable fields (e.g., `Handler`)
- Parity checks via `VerifyValidationEquivalence` and `VerifyUpdateValidationEquivalence`.

## Does this PR introduce a user-facing change?

```release-note
NONE

darshansreenivas commented:
/ok-to-test

yongruilin reviewed (commented):
Add the `// +k8s:validation-gen=xxx` tag for v1alpha1 as well?

darshansreenivas committed:
DRA: adding tag to node module

move DRA to a update method

fix boilerplate code

apply node normalize code at fuzz logic

adding normalization fix

fix nit

darshansreenivas commented:
/remove-sig apps
/remove-sig architecture
/remove-sig auth
/remove-sig autoscaling
/remove-sig cli
/remove-sig cloud-provider
/remove-sig cluster-lifecycle
/remove-sig etcd
/remove-sig instrumentation
/remove-sig network
/remove-sig release
/remove-sig scheduling
/remove-sig storage
/remove-sig testing
/remove-sig windows

lalitc375 reviewed (commented):
One comment, otherwise looks good.

darshansreenivas commented on pkg/api/testing/validation_test.go:
```diff
@@ -62,7 +63,8 @@ func TestVersionedValidationByFuzzing(t *testing.T) {
 					var opts []ValidationTestConfig
-					opts = append(opts, WithNormalizationRules(resourcevalidation.ResourceNormalizationRules...))
+					opts = append(opts, WithNormalizationRules(resourcevalidation.ResourceNormalizationRules...),
+						WithNormalizationRules(nodevalidation.NodeNormalizationRules...))
```
I see, sure, I will add rules in the constructor as well.

lalitc375 reviewed (commented):
Looks good,

two nits.

lalitc375 commented on pkg/api/testing/validation_test.go:
```diff
@@ -59,7 +64,12 @@ func TestVersionedValidationByFuzzing(t *testing.T) {
+					// TODO(API group level configuration): Consider configuring normalization rules at the
+					// API group level to avoid potential collisions when multiple rule sets are combined.
+					// This would allow each API group to register its own normalization rules independently.
+					allRules := append([]field.NormalizationRule{}, resourcevalidation.ResourceNormalizationRules...)
```
This append can be dropped, allRules can be directly instantiated using resourcevalidation.ResourceNormalizationRules.

lalitc375 reviewed (commented):
/lgtm

@thockin  Can you do the API review?

thockin commented on pkg/registry/node/runtimeclass/declarative_validation_test.go:
```diff
@@ -0,0 +1,130 @@
+					}),
+					expectedErrs: field.ErrorList{},
+				},
+				"invalid handler dns label": {
```
I think we should aim for better tests as we do the conversions.  Don't replicate every case of the DNS tests, but do test empty string, too-long, and ill-formed (e.g. uppercase)

lalitc375 replied:
I missed it, we should more invalid test cases here.

darshansreenivas replied:
@thockin and @lalitc375 added new test cases.

thockin commented on staging/src/k8s.io/api/node/v1/types.go:
```diff
@@ -50,6 +50,8 @@ type RuntimeClass struct {
 	// The Handler must be lowercase, conform to the DNS Label (RFC 1123) requirements,
 	// and is immutable.
+	// +k8s:format="k8s-short-name"
+	// +k8s:immutable
```
There's a push to make all fields either required or optional - I assume this should be required?

thockin reviewed (commented):
/approve for API

Please discuss with @lalitc375 about test cases and `required` if you think I am not right :)

Lalit has the LGTM.

darshansreenivas committed:
adding required and other test cases

darshansreenivas committed:
fixed order of tags

lalitc375 reviewed (commented):
Thank you

/lgtm

matthyx commented:
/retest
